File: springboot-backend/src/main/java/com/investa/service/AuthService.java
package com.investa.service;

import com.investa.dto.*;
import com.investa.model.User;
import com.investa.repository.UserRepository;
import com.investa.security.JwtUtil;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Slf4j
@Service
public class AuthService {
    private final UserRepository userRepository;
    private final JwtUtil jwtUtil;
    private final PasswordEncoder passwordEncoder;

    public AuthService(
            UserRepository userRepository,
            JwtUtil jwtUtil,
            PasswordEncoder passwordEncoder
    ) {
        this.userRepository = userRepository;
        this.jwtUtil = jwtUtil;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * Login user and return JWT tokens
     */
    public AuthResponse login(LoginRequest request) {
        log.info("Login attempt for email: {}", request.getEmail());

        // Find user by email
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Invalid email or password"));

        // Check if user is active
        if (!user.isActive()) {
            throw new RuntimeException("User account is inactive");
        }

        // Validate password
        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new RuntimeException("Invalid email or password");
        }

        // Update last login
        user.setLastLogin(LocalDateTime.now());
        userRepository.save(user);

        // Generate tokens
        String accessToken = jwtUtil.generateToken(user.getId(), user.getEmail());
        String refreshToken = jwtUtil.generateRefreshToken(user.getId(), user.getEmail());

        log.info("User logged in successfully: {}", user.getEmail());

        return AuthResponse.builder()
                .token(accessToken)
                .refreshToken(refreshToken)
                .user(mapToDTO(user))
                .message("Login successful")
                .build();
    }

    /**
     * Register new user
     */
    public AuthResponse register(RegisterRequest request) {
        log.info("Registration attempt for email: {}", request.getEmail());

        // Check if email already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already registered");
        }

        // Validate password confirmation
        if (!request.getPassword().equals(request.getConfirmPassword())) {
            throw new RuntimeException("Passwords do not match");
        }

        // Create new user
        User user = User.builder()
                .email(request.getEmail())
                .name(request.getName())
                .password(passwordEncoder.encode(request.getPassword()))
                .role("USER")
                .active(true)
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        User savedUser = userRepository.save(user);
        log.info("User registered successfully: {}", savedUser.getEmail());

        // Generate tokens
        String accessToken = jwtUtil.generateToken(savedUser.getId(), savedUser.getEmail());
        String refreshToken = jwtUtil.generateRefreshToken(savedUser.getId(), savedUser.getEmail());

        return AuthResponse.builder()
                .token(accessToken)
                .refreshToken(refreshToken)
                .user(mapToDTO(savedUser))
                .message("Registration successful")
                .build();
    }

    /**
     * Refresh access token using refresh token
     */
    public AuthResponse refreshToken(RefreshTokenRequest request) {
        log.info("Token refresh attempt");

        String refreshToken = request.getRefreshToken();

        // Validate refresh token
        if (!jwtUtil.isTokenValid(refreshToken)) {
            throw new RuntimeException("Invalid refresh token");
        }

        // Check token type
        String tokenType = jwtUtil.getTokenType(refreshToken);
        if (!"refresh".equals(tokenType)) {
            throw new RuntimeException("Invalid token type");
        }

        // Extract user ID and email
        String userId = jwtUtil.extractUserId(refreshToken);
        String email = jwtUtil.extractEmail(refreshToken);

        // Find user
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Check if user is active
        if (!user.isActive()) {
            throw new RuntimeException("User account is inactive");
        }

        // Generate new access token
        String newAccessToken = jwtUtil.generateToken(userId, email);
        String newRefreshToken = jwtUtil.generateRefreshToken(userId, email);

        log.info("Token refreshed successfully for user: {}", email);

        return AuthResponse.builder()
                .token(newAccessToken)
                .refreshToken(newRefreshToken)
                .user(mapToDTO(user))
                .message("Token refreshed successfully")
                .build();
    }

    /**
     * Get current user info
     */
    public UserDTO getCurrentUser(String userId) {
        log.info("Fetching current user: {}", userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        return mapToDTO(user);
    }

    /**
     * Map User entity to DTO
     */
    private UserDTO mapToDTO(User user) {
        return UserDTO.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole())
                .createdAt(user.getCreatedAt())
                .build();
    }
}

File: Investment Research Agent/src/services/auth-service.ts
import { apiClient } from "@/lib/api-client";
import type { AuthSession, User } from "@/types";

/**
 * Authentication service - connects to Spring Boot backend
 * Endpoints:
 *   POST /api/auth/login     -> { token, refreshToken, user }
 *   POST /api/auth/register  -> { token, refreshToken, user }
 *   POST /api/auth/refresh   -> { token, refreshToken, user }
 *   GET  /api/auth/me        -> { id, name, email, role, createdAt }
 *   GET  /api/auth/health    -> Health check
 */

interface AuthResponse {
  token: string;
  refreshToken: string;
  user: User;
  message?: string;
}

export const authService = {
  /**
   * Login with email and password
   */
  async login(email: string, password: string, _remember?: boolean): Promise<AuthSession> {
    try {
      const response = await apiClient.post<AuthResponse>("/auth/login", {
        email,
        password,
        remember: _remember,
      });

      if (!response.data.token || !response.data.user) {
        throw new Error("Invalid response from server");
      }

      return {
        token: response.data.token,
        refreshToken: response.data.refreshToken,
        user: response.data.user,
      };
    } catch (error: any) {
      const message = error.response?.data?.message || error.message || "Login failed";
      throw new Error(message);
    }
  },

  /**
   * Register new user
   */
  async register(name: string, email: string, password: string): Promise<AuthSession> {
    try {
      const response = await apiClient.post<AuthResponse>("/auth/register", {
        name,
        email,
        password,
        confirmPassword: password,
      });

      if (!response.data.token || !response.data.user) {
        throw new Error("Invalid response from server");
      }

      return {
        token: response.data.token,
        refreshToken: response.data.refreshToken,
        user: response.data.user,
      };
    } catch (error: any) {
      const message = error.response?.data?.message || error.message || "Registration failed";
      throw new Error(message);
    }
  },

  /**
   * Refresh access token
   */
  async refreshAccessToken(refreshToken: string): Promise<AuthSession> {
    try {
      const response = await apiClient.post<AuthResponse>("/auth/refresh", {
        refreshToken,
      });

      if (!response.data.token || !response.data.user) {
        throw new Error("Invalid response from server");
      }

      return {
        token: response.data.token,
        refreshToken: response.data.refreshToken,
        user: response.data.user,
      };
    } catch (error: any) {
      const message = error.response?.data?.message || error.message || "Token refresh failed";
      throw new Error(message);
    }
  },

  /**
   * Get current user info
   * Note: Requires valid JWT token in Authorization header
   */
  async me(): Promise<User> {
    try {
      const userId = localStorage.getItem("auth.user.id");
      if (!userId) {
        throw new Error("User ID not found");
      }

      const response = await apiClient.get<User>("/auth/me", {
        headers: {
          "X-User-Id": userId,
        },
      });

      return response.data;
    } catch (error: any) {
      const message = error.response?.data?.message || error.message || "Failed to fetch user";
      throw new Error(message);
    }
  },

  /**
   * Health check
   */
  async health(): Promise<boolean> {
    try {
      await apiClient.get("/auth/health");
      return true;
    } catch {
      return false;
    }
  },
};
